bin/istio-proxy-cfg.py

```diff
@@ -155,54 +162,101 @@ def pod_info():
+            print "podname must be either name,namespace,podip or name.namespace or any string that's a pod's label or a prefix of a pod's name"
             return None
 
-        return POD(si[0], si[1], si[2], None)
+        podname = si[0]
+        podns = si[1]
+        podip = si[2]
 
+    pods = []
     for pn, pod in pi.items():
-        if pn == searchstr:
-            return pod
-        if searchstr == pod.name:
-            return pod
-        if searchstr in pod.labels.values():
-            return pod
-
-    return None
+        if podname and podns:
+            if podip:
+                if podname == pod.name and podns == pod.namespace and podip == pod.ip:
+                    pods.append(pod)
+                    return pods
+            elif podname == pod.name and podns == pod.namespace:
+                pods.append(pod)
+                return pods
+        elif searchstr in pod.labels.values():
+            pods.append(pod)
+        elif pn.startswith(searchstr):
+            pods.append(pod)
+
+    return pods
+
+
+def find_pilot_url():
+    try:
+        pilot_svc = subprocess.check_output(
+            "kubectl get svc {svc} -n {ns} -o json".format(svc=PILOT_SVC, ns=ISTIO_NS).split())
+    except:
+        pilot_svc = {}
+    pilot_url = ""
+    if pilot_svc:
+        pilot_spec = json.loads(pilot_svc)['spec']
+        for port in pilot_spec['ports']:
+            if port['name'] == 'http-discovery':
+                pilot_url = "http://{ip}:{port}".format(ip=pilot_spec['clusterIP'], port=port['port'])
+                break
+    return pilot_url
 
 
 def main(args):
-    pod = searchpod(pod_info(), args.podname)
-    if pod is None:
+    pods = searchpod(pod_info(), args.podname)
+
+    if not pods:
         print "Cound not find pod ", args.podname
         return -1
 
-    if args.pilot_url:
-        print "Fetching from Pilot"
-        src = "pilot"
-        xds = XDS(url=args.pilot_url)
-        data = xds.lds(pod, True)
-    else:
-        print "Fetching from Envoy admin port via kubectl"
-        src = "proxy"
-        pr = Proxy(pod)
-        data = pr.routes()
+    if len(pods) > 1:
+        podnames = ["%s.%s" % (pod.name, pod.namespace) for pod in pods]
+        print "More than one pod is found: %s" % ", ".join(podnames)
+        return -1
+
+    pod = pods[0]
+    pilot_url = args.pilot_url
+    if not pilot_url:
+        pilot_url = find_pilot_url()
 
     if args.output is None:
-        args.output = pod.name + "_" + src + "_xds.yaml"
+        args.output = pod.name + "_xds.yaml"
 
     op = open(args.output, "wt")
+
+    op.write("==== Fetching from Pilot for pod %s in %s namespace\n" % (pod.name, pod.namespace))
+    print "Fetching from Pilot for pod %s in %s namespace" % (pod.name, pod.namespace)
+    xds = XDS(url=pilot_url)
+    data = xds.lds(pod, True)
     yaml.safe_dump(data, op, default_flow_style=False,
                    allow_unicode=False, indent=2)
+
+    op.write("\n\n")
```

Comment:
We should continue separating yaml files by source(pilot or proxy). It is easy to diff 2 separate files.

Reply:
I agree. I was thinking about that while working on it, but hesitating on what command arguments should be used to achieve that.  Before this PR, --pilot_url is used to indicate pilot vs proxy. Now that pilot_url is not required, and there are three types of output: pilot, proxy, cache_stats, need a way to write them in different files. Is it ok to define the output argument as --output [DIR]? That way, three files will be written in the directory. Do you have any suggestions or preference? Do you think it's necessary for user to indicate the type of output in the command line?

Reply:
Yes output dir sounds good